### ISSUE TITLE
Changes to the procedure order form. Currently it is just a lab order form, but other procedures should be allowed.

### ISSUE THREAD

**Issue by teryhill** [LibreHealthIO/LibreEHR#967]
Currently the procedure order form will send information to any Labs that are setup to Receive orders from the clinic. There is a dropdown at the bottom (which will need to be moved to the top of the form) of the form in which you could tag the procedure as some thin else. If any thing in the bottom section is chosen besides procedure or laboratory test there needs to be a print button. The field that is called procedure would need to look for the information as a different data type in the procedure codes table. 

![image](https://user-images.githubusercontent.com/4655366/35411328-5ab71fe8-01e6-11e8-979b-ffa4a43b36cd.png)

**Comment by tmccormi** [comment]
Add some controls to the Orders List that designate the type as "manual/electronic/both" to control whether the there is a Save/Transmit  Save/Print button visible.

**Comment by tmccormi** [comment]
I would rename this Form to be "Orders" in general, not Procedure Orders.

**Comment by aethelwulffe** [comment]
Agree on all points in this issue.

**Comment by teryhill** [comment]
Changes here to accommodate the electronic or paper options. Add two check boxes one for paper and one for electronic. This will allow for one or the other and both.

![image](https://user-images.githubusercontent.com/4655366/35414488-9969b48a-01f0-11e8-96b6-3135e20d7680.png)

**Comment by aethelwulffe** [comment]
Counseling, Psych eval....onandonandon.

**Comment by teryhill** [comment]
Changes so far

![image](https://user-images.githubusercontent.com/4655366/35414970-129b5754-01f2-11e8-8a71-be8465fe23b5.png)

**Comment by htuck** [comment]
Re: documenting the processing of a provider's order, I believe that everybody who touches it has to put their name to it.  I.e., the ordering provider has to sign it, and the transcribing staff has to sign that it was transcribed.  See the ovals in the pic below of the prov Order display in the encounter summary; is the encounter form's 'procedure order by' name legally equivalent to an e-sig?  If so, can have the transcribing staff e-sig the Provider Order.  Else will need the provider to e-sig the order form itself, and have the transcriber e-sign somewhere else that it was transcribed.

Would it be hard to add another e-sig button to the orfer form, saying 'transcribed'?  I have no idea what other things are touched by an e-sig.
Incidentally, how has this requirement been dealt with in the past, needing legally binding documentation of provider orders having been transcribed?

In other news, the results field (rectangle) will need to be cut out for non- lab orders.

![procordby](https://user-images.githubusercontent.com/8339489/35414975-f2c9d3ce-01d8-11e8-9180-0caa792703ad.png)

**Comment by aethelwulffe** [comment]
Since we are having the discussion here, I will end my connection to the Zoom meeting!
Hi Harley!

**Comment by htuck** [comment]
Hi Art!  Missed ya!

**Comment by teryhill** [comment]
The code that was committed was done for the training video.

Current needs:

The print button needs to be coded to work, the appearing of the print button needs to be driven by the toggle check boxes in the list_options table. 

The items that Harley describes in this issue needs to be addressed.

Future needs:

A means of displaying items from the address book as the send to option if the transmit option is selected and the option is not lab or procedure. If the option is imaging then the items in the address book with an imaging type need to display.

The procedure look up needs to be configurable to allow for other sets of data.

**Comment by aethelwulffe** [comment]
Keeping the issue open.

**Comment by 2208Abhinav** [comment]
Till now I have understood one issue that we need to put print button according to the field selected (I think we can use some javascript here :thinking: ), but I don't get the issue of **eSign** can someone explain this to me please.

**Comment by 2208Abhinav** [comment]
@teryhill I would like to work on this issue with you :smiley:

**Comment by teryhill** [comment]
Make sure you get the latest changes.

Esign is taken care of by the system. This is where the Dr electronically signs the form saying it is a legal document. We may need to add a for for transcription or other things but for now we just need to sart with this form.

![image](https://user-images.githubusercontent.com/4655366/35472803-0989cf1c-0344-11e8-864c-1b9eed5e84ed.png)

**Comment by 2208Abhinav** [comment]
Thanks, right now I am working on the print button part.

**Comment by aethelwulffe** [comment]
Please refresh your branch against master as you proceed.  There may have been some related changes.

**Comment by 2208Abhinav** [comment]
I did it thank you @aethelwulffe .

**Comment by 2208Abhinav** [comment]
There is a serious issue @aethelwulffe .When I updated my master branch and tried to work on it after restarting nothing appeared on the login.php page.

**Comment by teryhill** [comment]
@2208Abhinav look at the php error log and see if it shows an error. Clear your Cache.

**Comment by 2208Abhinav** [comment]
Its showing **syntax error, unexpected end of file in /opt/lampp/htdocs/LibreEHR/interface/globals.php on line 538**
I am searching for the error.

**Comment by 2208Abhinav** [comment]
Now I remember one quote which I saw when I was studying php -

> Debugging is twice as hard as writing the code in the first place.Therefore, if you write the code as cleverly as possible, you are, by definition, not smart enough to debug it.
:smile:

**Comment by 2208Abhinav** [comment]
I worked out for the print button part and its working fine.
I added `id="print_button"` for **Print** button and used the code below
![screenshot from 2018-01-28 20-43-34](https://user-images.githubusercontent.com/30211121/35480881-1d57d0b6-043e-11e8-9c8b-90939aede1e2.png)

**Comment by teryhill** [comment]
@2208Abhinav look at the table for list_options. There are 2 fields toggle 1 and toggle 2. If you look at the edit list you will see that they say paper and electronic. Code this using those fields if the one for paper is checked show the print button.

**Comment by 2208Abhinav** [comment]
@teryhill where we find the options to select between paper and electronic?

**Comment by teryhill** [comment]
@2208Abhinav go to Administration then lists. Then look for order types and you should see them

![image](https://user-images.githubusercontent.com/4655366/35625954-26d69caa-0662-11e8-9006-a6ef5fa5db72.png)

**Comment by 2208Abhinav** [comment]
So now I am required to put two checks
First for selection between **Paper** and **Electronic**
Second for selection between **Procedures**
If **Paper** is chosen in the first check then go for second check otherwise its not needed, remove **Print** button from there for all **Procedures** if **Electronic** is chosen.

**Comment by teryhill** [comment]
Just show the "Print" button if paper is selected no matter what the title is. If electronic and paper is selected show the print button. If electronic is the only choice do not sow  the print button. The type of procedure should not matter in theory you could send an image order request electronically or on paper.

**Comment by 2208Abhinav** [comment]
I got it, **Print** is just affected by **Electronic** or **Paper** nothing else.
If **Paper** is chosen or both **Electronic** and **Paper** are chosen then show **Print**.
Else don't.

**Comment by teryhill** [comment]
Yes you understand.

**Comment by htuck** [comment]
I have the latest code from this am and it is all very cool, guys, great work!  
As far as I can tell, the "Orders" form is ready to mock up that part of the video.
Signing off the order transcription (i.e., the e-signing issue) seems to be a different concern from all this- shall I add a new issue for that?

**Comment by teryhill** [comment]
@htuck yes add an issue for the transcription form. We will have to design on and discuss how to accomplish it. If it is a form it should e-sign. I think we just need the form.

**Comment by teryhill** [comment]
@tmccormi can you setup @2208Abhinav so I can assign this task or will you assign him this task.

**Comment by htuck** [comment]
Curious I am what happens when you click the Print button on the newly repurposed Orders form when the Order Type is not 'Procedure Order'.

Does it print out a referral form, such as you'd get from the Transactions module (pt summary screen, Transactions link along the top) if a referral were being sent out?  I'm trying to avoid setting up a whole fricking Procedure Order to test this thing, but it may be inevitable.

Seems like we're trying to make the new mods to the Procedure Order form serve two masters: its original old- school Proc Ord module, and now the Transactions module where we would output referrals for Order Types other than "Procedure Orders".

If the outputs of the 'new' Orders form get put into the same accountability stacks as the Transaction documents, that's fine.  But if they're separate because they were generated by the Proc Ord form, wouldn't that be a problem?

**Comment by aethelwulffe** [comment]
I sent an invite for him to be a collaborator.

**Comment by teryhill** [comment]
@aethelwulffe do you think that should be something I could do?

**Comment by aethelwulffe** [comment]
@teryhill yes.  Permission settings for the whole org seem easier to change than per-repo. You know have ultimate super-powers.  Castle GreySkull and all that.

**Comment by aethelwulffe** [comment]
Found where I can do the changes on a granular scale, but frankly I don't think anything less than org superadmin is really appropriate.

**Comment by teryhill** [comment]
Thanks Orko

**Comment by aethelwulffe** [comment]
Had to look that up...
There is a certain resemblance.

**Comment by teryhill** [comment]
@2208Abhinav what is your user on the chat?

**Comment by 2208Abhinav** [comment]
@teryhill I have pulled a request #988 for **PRINT** button visibility according to the selection of **PAPER**, please review it.

**Comment by 2208Abhinav** [comment]
@teryhill are we over with print button part here?

**Comment by teryhill** [comment]
@2208Abhinav  yes the print feature is working for the print button.

**Comment by teryhill** [comment]
@2208Abhinav the standard report and the code to print it or another report if it exists would be next, If you are willing to continue.

**Comment by 2208Abhinav** [comment]
@teryhill I am ready please send me the details.

**Comment by teryhill** [comment]
@2208Abhinav Look at the PDF I sent you on the forums. That will be the standard report.

You will need to do a check and see if a file with the same name as the list option exists, if it does use that for the print file else use the standard printout.

**Comment by teryhill** [comment]
The insurance function that was added to patient .inc will give you the insurance information. The facility information is in facility and the actual facility used in in the form encounter table. You will need 2 primary pieces of data to start the report PID and encounter number.

**Comment by teryhill** [comment]
$insurance = getAllinsurances($pid);
Will get the insurance information

**Comment by htuck** [comment]
The Procedure Order form has a couple problematic text areas which I haven't seen any mention of having been addressed.

	'status' - derives list contents ONLY from the old- school configured Procedure Orders- can it be disappeared if order type is not 'Procedure Order'?
	'sending to': still needs to take provider list from the addr book
	'Procedure' ddl - takes from the configured Procedure Order list of configured labs or test organizations - can it be disappear if order type is not 'Procedure Order'?
		For the new Provider Orders that info will be provided in the Test Description

**Comment by teryhill** [comment]
Sending to currently take the entries from the addr book as long as the entry has a "Lab Service" type. This will be changed to something else like maybe Ordering Facility.

Procedure can be setup for anything. I set it up in the files I sent for an imaging location. It could be removed if the status was not procedure.

Status I am not sure what the thought process was for this, but any order could be canceled, Completed , perding or routed (to someone else). This could be removed if need be.

**Comment by htuck** [comment]
Yabbut if a clinic for some reason wants to config and use the Procedure Order module, they'll need the 'Procedure' field with all the code that powers it.  It looks to me like Test Description serves the same purpose for the 'Provider Orders', which is why I was thinking to disappear, or grey out, the 'Procedure' field if Order Type <> Procedure Order.

The old Procedure Order module has a 'status' column in the  'Pending Review' and 'Patient Results' screens.  That value automatically updates as the Procedure Order progresses from being ordered, reviewed and noted.  So that tells me that dropdown feeds or is populated by something that's pretty involved in the Proc Ord code which is why I thought it should be treated like the 'Procedure' field above.

If it would be easy to make 'Sending to' simply read all the providers in the Addr book, like the Transactions panel's "Refer To" does, that would solve that problem.

**Comment by 2208Abhinav** [comment]
@teryhill are you working on the code to create printouts?

**Comment by teryhill** [comment]
@2208Abhinav no I am not. I have not had much time to devote to writing code for this issue.

**Comment by 2208Abhinav** [comment]
Busy in GSOC stuff?

**Comment by teryhill** [comment]
And work 8^)

**Comment by 2208Abhinav** [comment]
Most important

**Comment by teryhill** [comment]
You can do the report You are talented and that should be easy

**Comment by teryhill** [comment]
I need to go for a while

**Comment by 2208Abhinav** [comment]
I was thinking of using TCPDF for pdf generation instead of html2pdf

**Comment by aethelwulffe** [comment]
@2208Abhinav If TCPDF is a better option, and since we are trying to replace html2pdf, I support that effort.

**Comment by 2208Abhinav** [comment]
After some further research I found that FPDF is much better than TCPDF.
FPDF is 7-10 times faster, PHP source file of FPDF is 26 times smaller than TCPDF.
FPDF v1.8(2015-11-29) is the latest version available.

**Comment by teryhill** [comment]
But is TCPDF being actively updated It does not look like FPDF is with a version date of 2 years ago.

**Comment by aethelwulffe** [comment]
That is a concern...and I think we have or had FPDF in the system somewhere.

**Comment by aethelwulffe** [comment]
Of course, if it has not been changed in years because it is perfect, that is another matter.  I dunno.

**Comment by 2208Abhinav** [comment]
@aethelwulffe we are using TCPDF, FPDF, html2pdf and others maybe. My aim is to remove many libraries and use one.

**Comment by 2208Abhinav** [comment]
I think FPDF is serving well thats why its still in demand without any update for 2 years.
